Allow to provide custom configuration for JAXB context

Sgitario opened the pull request:
Changes added as part of this pull request:

When using the `quarkus-resteasy-reactive-jaxb` extension there are some advanced features that RESTEasy Reactive supports.

1. Inject JAXB components

The JAXB resteasy reactive extension will serialize and deserialize requests and responses transparently for users. However, if you need finer usage of the JAXB components, you can inject either the JAXBContext, Marshaller, or Unmarshaller components into your beans:

```java
@ApplicationScoped
public class MyService {

    @Inject
    JAXBContext jaxbContext;

    @Inject
    Marshaller marshaller;

    @Inject
    Unmarshaller unmarshaller;

    // ...
}
```

Note: Quarkus will automatically register all the classes annotated with `@XmlRootElement` to the JAXB context.

2. Customize the JAXB configuration

To customize the JAXB configuration for either the JAXB context, and/or the Marshaller/Unmarshaller components, the suggested approach is to define a CDI bean of type `io.quarkus.jaxb.runtime.JaxbContextCustomizer`.

An example where a custom module needs to be registered would look like so:

```java
@Singleton
public class RegisterCustomModuleCustomizer implements JaxbContextCustomizer {

    // For JAXB context configuration
    @Override
    public void customizeContextProperties(Map<String, Object> properties) {

    }

    // For Marshaller configuration
    @Override
    public void customizeMarshaller(Marshaller marshaller) throws PropertyException {
        marshaller.setProperty("jaxb.formatted.output", Boolean.TRUE);
    }

    // For Unmarshaller configuration
    @Override
    public void customizeUnmarshaller(Unmarshaller unmarshaller) throws PropertyException {
        // ...
    }
}
```

Note: It's not necessary to implement the three methods, but only the want you need.

Alternatively, you can provide your own `JAXBContext` bean by doing:

```java
public class CustomJaxbContext {

    // Replaces the CDI producer for JAXBContext built into Quarkus
    @Singleton
    JAXBContext jaxbContext() {
        // ...
    }
}
```

Note: By replacing the JAXB context, you need to bound the classes to be serialized/deserialized and also configure the JAXB context accordingly. 

Resolves https://github.com/quarkusio/quarkus/issues/25753

geoand commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1138,6 +1138,92 @@ Importing this module will allow HTTP message bodies to be read from XML
+
+[NOTE]
+====
+Quarkus will automatically register all the classes annotated with `@XmlRootElement` to the JAXB context.
```
It is not clear what `automatically register` means in this context

Sgitario replied:
Replaced with:
`Quarkus will automatically find all the classes annotated with `@XmlRootElement` and then bound them to the JAXB context.`

geoand commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1138,6 +1138,92 @@ Importing this module will allow HTTP message bodies to be read from XML
+
+===== Inject JAXB components
+
+The JAXB resteasy reactive extension will serialize and unserialize requests and responses transparently for users. However, if you need finer usage of the JAXB components, you can inject either the JAXBContext, Marshaller, or Unmarshaller components into your beans:
```
```suggestion
The JAXB resteasy reactive extension will serialize and unserialize requests and responses transparently for users. However, if you need finer grain control over JAXB components, you can inject either the JAXBContext, Marshaller, or Unmarshaller components into your beans:
```

geoand commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1138,6 +1138,92 @@ Importing this module will allow HTTP message bodies to be read from XML
+
+[NOTE]
+====
+It's not necessary to implement the three methods, but only the want you need.
```
```suggestion
It's not necessary to implement all three methods, but only the want you need.
```

geoand commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1138,6 +1138,92 @@ Importing this module will allow HTTP message bodies to be read from XML
+public class CustomJaxbContext {
+
+    // Replaces the CDI producer for JAXBContext built into Quarkus
+    @Singleton
```
I would also add `@Produces` here

Sgitario replied:
+1
I've also updated the guide of `rest-json.adoc` where I took this code snippet from.

geoand commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1138,6 +1138,92 @@ Importing this module will allow HTTP message bodies to be read from XML
+
+[IMPORTANT]
+====
+By replacing the JAXB context, you need to bound the classes to be serialized/unserialized and also configure the JAXB context accordingly. 
```
> you need to bound the classes

What does this mean?

Sgitario replied:
Updated with:
`Note that if you provide your custom JAXB context instance, you will need to register the classes you want to use for the XML serialization. This means that Quarkus will not update your custom JAXB context instance with the auto-discovered classes. `

geoand commented on extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java:
```diff
@@ -303,8 +319,11 @@ void ignoreWarnings(BuildProducer<ReflectiveHierarchyIgnoreWarningBuildItem> ign
+            BuildProducer<ServiceProviderBuildItem> providerItem,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
+            BuildProducer<NativeImageResourceBundleBuildItem> resourceBundle) {
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, "com.sun.xml.bind.v2.ContextFactory"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, "com.sun.xml.internal.bind.v2.ContextFactory"));
```
Why are these now needed?

Sgitario replied:
This is due to a wrong merge on my side. Removed. Thanks for spotting!

geoand commented on extensions/jaxb/runtime/src/main/java/io/quarkus/jaxb/runtime/JaxbContextConfigRecorder.java:
```diff
@@ -0,0 +1,36 @@
+
+@Recorder
+public class JaxbContextConfigRecorder {
+    private static Set<Class<?>> classesToBeBound = new HashSet<>();
+    private static Set<String> stringClassesToBeBound = new HashSet<>();
```
These probably need to be `volatile`

Sgitario replied:
Added

geoand reviewed (commented):
Thanks for this!

I added an initial round of comments

Sgitario commented:
> Thanks for this!
> 
> I added an initial round of comments

Many thanks! 
I think I've already addressed all your comments. Plus, I've also deleted the `stringClassesToBeBound` from JaxbContextConfigRecorder.

geoand commented:
> > Thanks for this!
> > I added an initial round of comments
> 
> Many thanks! I think I've already addressed all your comments. Plus, I've also deleted the `stringClassesToBeBound` from JaxbContextConfigRecorder.

Great!

I'll take another look later.
@Postremus, @gsmet do you also want to have a look?

gastaldi commented on extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbClassesToBeBoundBuildItem.java:
```diff
@@ -0,0 +1,21 @@
+ */
+public final class JaxbClassesToBeBoundBuildItem extends MultiBuildItem {
+
+    private final List<Class<?>> classes;
```
Maybe store only the class names here (`List<String>`). Other extensions may have use cases where the `Class` isn't available

Sgitario replied:
I didn't want to do this at first because we will be loading the classes from the class names at runtime (so, the startup time will be degraded). But perhaps it will be insignificant and using a list of strings is more usable. 
So, changed following your suggestion.

gastaldi replied:
Yes, I agree that it should be insignificant (Depending on the number of classes though), but you can avoid reflection completely if you generate the producer class in a build step

gastaldi commented on extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java:
```diff
@@ -275,8 +292,35 @@ void registerClasses(
+        if (capabilities.isPresent(Capability.HIBERNATE_VALIDATOR)) {
+            Class<?> validationReportClass;
+            if (capabilities.isPresent(Capability.RESTEASY)) {
+                validationReportClass = org.jboss.resteasy.api.validation.ViolationReport.class;
```
This should be produced in the `RestEasy` extension

Sgitario replied:
Moved to `resteasy-jaxb`

gastaldi commented on extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java:
```diff
@@ -275,8 +292,35 @@ void registerClasses(
+            if (capabilities.isPresent(Capability.RESTEASY)) {
+                validationReportClass = org.jboss.resteasy.api.validation.ViolationReport.class;
+            } else {
+                validationReportClass = io.quarkus.hibernate.validator.runtime.jaxrs.ViolationReport.class;
```
This should be produced in the `hibernate-validator` extension

Sgitario replied:
Moved to `resteasy-reactive-jaxb` instead of the `hibernate-validator` extension.

gastaldi commented on extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbClassesToBeBoundBuildItem.java:
```diff
@@ -0,0 +1,21 @@
+    private final List<Class<?>> classes;
+
+    public JaxbClassesToBeBoundBuildItem(List<Class<?>> classes) {
+        this.classes = classes;
```
```suggestion
        this.classes = Objects.requireNonNull(classes);
```

Sgitario replied:
Changed

gastaldi commented:
I like the idea, we should mark this as a noteworthy feature

gastaldi commented on extensions/jaxb/deployment/pom.xml:
```diff
@@ -25,6 +25,48 @@
+
+        <dependency>
+            <groupId>org.jboss.resteasy</groupId>
+            <artifactId>resteasy-core</artifactId>
```
Is this dependency still needed?

Sgitario replied:
Ops, thanks for spotting!

gastaldi commented on extensions/jaxb/runtime/src/main/java/io/quarkus/jaxb/runtime/JaxbContextConfigRecorder.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public static String[] getClassesToBeBound() {
+        return classesToBeBound.toArray(new String[classesToBeBound.size()]);
```
```suggestion
        return classesToBeBound.toArray(new String[0]);
```

gastaldi replied:
I think you can have a method here returning Class[] since that's what's going to be used when invoking `JAXBcontext.newInstance`

Sgitario replied:
I prefer doing the mapping from string to class in the `JAXBContext.newInstance` method.

gastaldi commented on extensions/jaxb/runtime/src/main/java/io/quarkus/jaxb/runtime/JaxbContextProducer.java:
```diff
@@ -0,0 +1,93 @@
+                    classes.add(clazz);
+                }
+            }
+            return JAXBContext.newInstance(classes.toArray(new Class[classes.size()]), properties);
```
```suggestion
            return JAXBContext.newInstance(classes.toArray(new Class[0]), properties);
```

gastaldi reviewed (approved):
LGTM

gastaldi commented:
Apparently the build is failing: 
```
2022-06-08 14:50:46,826 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-0) Request failed: java.lang.RuntimeException: javax.xml.bind.JAXBException: io.quarkus.resteasy.reactive.jaxb.deployment.test.SimpleXmlTest$ModelWithoutAnnotation is not known to this context
	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.unmarshal(JaxbMessageBodyReader.java:67)
	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.doReadFrom(JaxbMessageBodyReader.java:76)
	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.readFrom(JaxbMessageBodyReader.java:36)
	at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.readFrom(RequestDeserializeHandler.java:107)
	at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:67)
```

Sgitario committed:
Allow to provide custom configuration for JAXB context

==== Advanced JAXB-specific features

When using the `quarkus-resteasy-reactive-jaxb` extension there are some advanced features that RESTEasy Reactive supports.

===== Inject JAXB components

The JAXB resteasy reactive extension will serialize and unserialize requests and responses transparently for users. However, if you need finer usage of the JAXB components, you can inject either the JAXBContext, Marshaller, or Unmarshaller components into your beans:

[source,java]
----
@ApplicationScoped
public class MyService {

    @Inject
    JAXBContext jaxbContext;

    @Inject
    Marshaller marshaller;

    @Inject
    Unmarshaller unmarshaller;

    // ...
}
----

[NOTE]
====
Quarkus will automatically register all the classes annotated with `@XmlRootElement` to the JAXB context.
====

===== Customize the JAXB configuration

To customize the JAXB configuration for either the JAXB context, and/or the Marshaller/Unmarshaller components, the suggested approach is to define a CDI bean of type `io.quarkus.jaxb.runtime.JaxbContextCustomizer`.

An example where a custom module needs to be registered would look like so:

[source,java]
----
@Singleton
public class RegisterCustomModuleCustomizer implements JaxbContextCustomizer {

    // For JAXB context configuration
    @Override
    public void customizeContextProperties(Map<String, Object> properties) {

    }

    // For Marshaller configuration
    @Override
    public void customizeMarshaller(Marshaller marshaller) throws PropertyException {
        marshaller.setProperty("jaxb.formatted.output", Boolean.TRUE);
    }

    // For Unmarshaller configuration
    @Override
    public void customizeUnmarshaller(Unmarshaller unmarshaller) throws PropertyException {
        // ...
    }
}
----

[NOTE]
====
It's not necessary to implement the three methods, but only the want you need.
====

Alternatively, you can provide your own `JAXBContext` bean by doing:

[source,java]
----
public class CustomJaxbContext {

    // Replaces the CDI producer for JAXBContext built into Quarkus
    @Singleton
    JAXBContext jaxbContext() {
        // ...
    }
}
----

[IMPORTANT]
====
By replacing the JAXB context, you need to bound the classes to be serialized/unserialized and also configure the JAXB context accordingly.
====

Sgitario commented:
> Apparently the build is failing:
> 
> ```
> 2022-06-08 14:50:46,826 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-0) Request failed: java.lang.RuntimeException: javax.xml.bind.JAXBException: io.quarkus.resteasy.reactive.jaxb.deployment.test.SimpleXmlTest$ModelWithoutAnnotation is not known to this context
> 	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.unmarshal(JaxbMessageBodyReader.java:67)
> 	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.doReadFrom(JaxbMessageBodyReader.java:76)
> 	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.readFrom(JaxbMessageBodyReader.java:36)
> 	at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.readFrom(RequestDeserializeHandler.java:107)
> 	at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:67)
> ```

Yep... This is a sneaky test... It could not reproduce it locally, so I'm adding traces just to see what is going on.

Sgitario commented:
> > Apparently the build is failing:
> > ```
> > 2022-06-08 14:50:46,826 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-0) Request failed: java.lang.RuntimeException: javax.xml.bind.JAXBException: io.quarkus.resteasy.reactive.jaxb.deployment.test.SimpleXmlTest$ModelWithoutAnnotation is not known to this context
> > 	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.unmarshal(JaxbMessageBodyReader.java:67)
> > 	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.doReadFrom(JaxbMessageBodyReader.java:76)
> > 	at io.quarkus.resteasy.reactive.jaxb.common.runtime.serialisers.JaxbMessageBodyReader.readFrom(JaxbMessageBodyReader.java:36)
> > 	at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.readFrom(RequestDeserializeHandler.java:107)
> > 	at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:67)
> > ```
> 
> Yep... This is a sneaky test... It could not reproduce it locally, so I'm adding traces just to see what is going on.

Maybe it was failing due to some other inconsistencies, but now it seems to be working all the time.

geoand commented:
So is this ready or not?

ylepikhov commented:
Hi! How about to provide a way to disable this feature via configuration? Or maybe separate it from core jaxb support in additional jar? Point is that someone (as I, for example) may have it's own code producing JAXBContext which may (or may not) conflict with this feature. Second point is injecting Marshaller/Unmarshaller in @ApplicationScoped beans may result in errors because Marshaller/Unmarshaller are not thread safe.

geoand commented:
The `JAXBContext`, `Marshaller` and `Unmarshaller` beans are annotated with `@DefaultBean`, meaning that any user supplied beans will override what Quarkus provides. 

> Second point is injecting Marshaller/Unmarshaller in @ApplicationScoped beans may result in errors because Marshaller/Unmarshaller are not thread safe

@Sgitario I think we need to make these beans `@RequestScoped`... It may account for the flaky tests you have seen

geoand commented:
https://github.com/quarkusio/quarkus/pull/26231 should make the necessary changes
